pkg/graveler/settings/manager.go

```diff
@@ -0,0 +1,125 @@
+// performance of a storage put/get respectively.
+type Manager struct {
+	refManager                  graveler.RefManager
+	branchLock                  graveler.BranchLocker
+	blockAdapter                block.Adapter
+	committedBlockStoragePrefix string
+}
+
+// Save persists the given proto.Message under the given repository and key.
+func (m *Manager) Save(ctx context.Context, repositoryID graveler.RepositoryID, key string, message proto.Message) error {
+	repo, err := m.refManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	messageBytes, err := proto.Marshal(message)
+	if err != nil {
+		return err
+	}
+	err = m.blockAdapter.Put(ctx, block.ObjectPointer{
+		StorageNamespace: string(repo.StorageNamespace),
+		Identifier:       fmt.Sprintf(settingsSuffixTemplate, m.committedBlockStoragePrefix, key),
+		IdentifierType:   block.IdentifierTypeRelative,
+	}, int64(len(messageBytes)), bytes.NewReader(messageBytes), block.PutOpts{})
+	if err != nil {
+		return err
+	}
+	logger := logging.FromContext(ctx)
+	if logger.IsTracing() {
+		logger.
+			WithField("repo", repositoryID).
+			WithField("key", key).
+			WithField("setting", protojson.Format(message)).
+			Trace("saving repository-level setting")
+	}
+	return nil
+}
+
+// Get fetches the configuration under the given repository and key.
+// The result is placed in message.
+func (m *Manager) Get(ctx context.Context, repositoryID graveler.RepositoryID, key string, message proto.Message) error {
+	repo, err := m.refManager.GetRepository(ctx, repositoryID)
+	if err != nil {
+		return err
+	}
+	objectPointer := block.ObjectPointer{
+		StorageNamespace: string(repo.StorageNamespace),
+		Identifier:       fmt.Sprintf(settingsSuffixTemplate, m.committedBlockStoragePrefix, key),
+		IdentifierType:   block.IdentifierTypeRelative,
+	}
+	reader, err := m.blockAdapter.Get(ctx, objectPointer, -1)
+	if errors.Is(err, adapter.ErrDataNotFound) {
+		return graveler.ErrNotFound
+	}
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = reader.Close()
+	}()
+	messageBytes, err := io.ReadAll(reader)
+	if err != nil {
+		return err
+	}
+	err = proto.Unmarshal(messageBytes, message)
+	if err != nil {
+		return err
+	}
+	logger := logging.FromContext(ctx)
+	if logger.IsTracing() {
+		logger.
+			WithField("repo", repositoryID).
+			WithField("key", key).
+			WithField("setting", protojson.Format(message)).
+			Trace("got repository-level setting")
+	}
+	return nil
+}
+
+// UpdateWithLock atomically performs the following sequence:
+// 1. Fetch the configuration under the given repository and key, and place it in message.
+// 2. Call the given update function (which should presumably update the message).
+// 3. Persist message to the store.
+func (m *Manager) UpdateWithLock(ctx context.Context, repositoryID graveler.RepositoryID, key string, message proto.Message, update func()) error {
```

Comment:
Getting the settings while locking is not very useful if you act upon the settings after releasing the lock (settings might already changed..).

Using the `update` for doing the operation you wanted is one way of solving it, but it leads to some limitations, like not using the branch locker. Aside from the abstraction leak, Not sure we want to switch our entire branching technique to use this settings manager instead.

Reply:
The lock here is for doing the update atomically, not to use it for the settings.
Not sure I understand the comment